packages/cozy-client/src/models/document/qualification.spec.js

```diff
@@ -133,6 +133,12 @@ describe('qualifications items', () => {
     })
   })
 
+  it('should never have undefined attribute', () => {
+    // This qualification has no sourceSubCategory attribute
+    const qualification = Qualification.getByLabel('other_tax_document')
+    expect(qualification.hasOwnProperty('sourceSubCategory')).toBe(false)
+  })
```

Comment:
We should not rely on that "side effect". Because if we change `other_tax_document` to have a sourceSubCaterogry, the test will fail. This is not what is attended. 

Also, you should not use `howOwnProperty` but `hasOwn` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn

Reply:
No, it's not. Unit test should not depends on side effect. 🙏

Reply:
Not specially. I don't think we need to map on all the qualificationModel, but why not... 

I would have done a mock of the qualificationModel to write what I want and that's all. 

Also, I don't know if the test is right anymore. Before you tested that the key was not present anymore, now you test that the value is not undefined. This is not the same thing. Are you sure about the choice you made?

Reply:
I have trouble to understand why it takes so much time and discussions :confused: 
To mock the qualification file, you can simply do: 
```
jest.mock('../../assets/qualifications.json', () => ({
  qualifications: [
    {
      label: 'isp_invoice',
      purpose: 'invoice',
      sourceCategory: 'telecom',
      sourceSubCategory: 'internet',
      subjects: ['subscription']
    },
    {
      label: 'national_id_card',
      purpose: 'attestation',
      sourceCategory: 'gov',
      sourceSubCategory: 'civil_registration',
      subjects: ['identity']
    },
    {
      label: 'health_invoice',
      purpose: 'invoice',
      sourceCategory: 'health'
    },
    {
      label: 'other_identity_document',
      purpose: 'attestation',
      subjects: ['identity']
    }
  ],
  purposeKnownValues: ['attestation', 'invoice'],
  sourceCategoryKnownValues: ['telecom', 'health', 'gov'],
  sourceSubCategoryKnownValues: ['civil_registration', 'internet'],
  subjectsKnownValues: ['identity', 'subscription']
}))
```

And if you just want to test there is no undefined attribute: 
```
  it('should not have undefined attributes', () => {
    const qualification = Qualification.getByLabel('other_identity_document')
    Object.keys(qualification).forEach(key => {
      expect(qualification[key]).toBeDefined()
    })
  })
  ```
  And it's done

Reply:
> And is the same as my first test about other_tax_document. No ?

Bah non, ici on mock le model. Donc si on change le vrai code du model, ça n'aura pas d'impact sur le test. En gros, si je rajoute un sourceSubCaterogry à `other_tax_document` ça va pas péter le test. Et c'est ce qu'on veut. Donc ce mock est très bien. 

Après le test, j'aurais plus vu un truc du genre : 
```
it('should not have undefined attributes', () => {
    const qualification = Qualification.getByLabel('other_identity_document')
    expect(qualification.hasOwn('sourceSubCategory')).toBe(false)
  })
```

Reply:
> That do not answer to Crash first request about code really working

I don't know what "code really working" means, but the point was to not rely on side-effect, i.e. the actual qualifications file. Here I mock the qualification file, so it answers this point.

> And is the same as my first test about other_tax_document. No ?

No, since your test was relying on the real qualifications file. It the real `other_tax_document` qualification change, your test might fail. If you mock the file as suggested, it won't, since the mocked values won't change.

Reply:
De ce que je vois, ça reprend ma proposition, mais pas celle de @Crash-- , donc pas sûr que ce soit une synthèse ? 
C'est ok de mon côté, mais il manque ce qui était suggéré avec `expect(qualification.hasOwn('sourceSubCategory')).toBe(false)`